### PR TITLE
use only sync code for postgres

### PR DESCRIPTION
**Pull request by vincentsarago**
Opposite to #2 

This PR removes any `async` calls to run only `sync` operations (and then let fastapi wraps the method in a thread)

Before this PR we would have Async (asyncpg) and blocking (rasterio/gdal) calls within `async` endpoints. This was blocking the even loop. In #2 I've tried to create `fake` async method wrapping all rasterio/gdal calls but I think it's easier to use `sync` methods for PG interaction and let FastAPI wraps the whole endpoint in a thread https://github.com/tiangolo/fastapi/blob/dc5a966548df973d0fd2d3a28bad65d0436f6fb9/fastapi/routing.py#L141-L151

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ services:
       - POSTGRES_PORT=5432
       - WEB_CONCURRENCY=10
       - DB_MIN_CONN_SIZE=1
-      - DB_MAX_CONN_SIZE=1
+      - DB_MAX_CONN_SIZE=10
```
Comment (vincentsarago) [inline docker-compose.yml:24]: when having only 1 conn I get `psycopg2.pool.PoolError: connection pool exhausted`

**Comment by vincentsarago** [comment]
@bitner could you help review this? 🙏

**Comment by vincentsarago** [comment]
if anyone could review my `psycopg2` code it would be really appreciated 🙏

File: docker-compose.yml (hunk 1 above)
Comment (bitner) [reply docker-compose.yml:24]: I think you need to have DB_MAX_CONN_SIZE at least the same size as WEB_CONCURRENCY

File: src/titiler/pgstac/titiler/pgstac/factory.py [hunk 2]
```diff
@@ -292,14 +291,16 @@ async def create_mosaic(
...
-                    """,
-                    req=body.json(exclude_none=True),
-                )
-                mosaicid = await conn.fetchval(q, *p)
+            conn = pool.getconn()
+            try:
+                with conn:
+                    with conn.cursor() as cursor:
+                        cursor.execute(
+                            f"SELECT * FROM create_mosaic('{body.json(exclude_none=True)}'::text::jsonb);"
```
Comment (bitner) [inline src/titiler/pgstac/titiler/pgstac/factory.py:299]: Better SQL Injection practice:
```
cursor.execute("""
      SELECT * FROM create_mosaic(%s::text::jsonb);
""", (body.json(exclude_none=True))
)

```

File: src/titiler/pgstac/titiler/pgstac/mosaic.py [hunk 3]
```diff
@@ -74,36 +73,38 @@ def _read(self) -> MosaicJSON:
...
-                mosaicid=self.path,
-                geom=geom.json(exclude_none=True),
-            )
-            items = await conn.fetchval(q, *p)
+        conn = self.pool.getconn()
+        try:
+            with conn:
+                with conn.cursor() as cursor:
+                    cursor.execute(
+                        f"SELECT * FROM items_for_geom('{self.path}'::text, '{geom.json(exclude_none=True)}'::text::jsonb);"
```
Comment (bitner) [inline src/titiler/pgstac/titiler/pgstac/mosaic.py:95]: Similarly:
```
cursor.execute("SELECT * FROM items_for_geom(%::text, %::text::jsonb);",(self.path,geom.json(exclude_none=True),))

**Review by bitner — CHANGES_REQUESTED** [review]
Couple things to clean up re SQL injection vulnerability.

**Comment by bitner** [comment]
I have to say the whole async blocking vs sync thing still makes my head swim a bit.

**Comment by vincentsarago** [comment]
> I have to say the whole async blocking vs sync thing still makes my head swim a bit.

@bitner, rasterio/gdal is a blocking process. if you do a blocking thing (deadlock) in an async call, you'll end up blocking your async. So it's better to run the code in a full `sync` way, which fastAPI will wrap in a thread (enabling the application to work like if the endpoint was async) instead of having a mix of sync/async process in a function. 

@geospatial-jeff could talk more about this ;-)

**Pull request merged**